### ISSUE TITLE
Efficiently loading Waymo raw data? 

### ISSUE THREAD

**Issue by aryasenna** [waymo-research/waymo-open-dataset#856]
Hello,

I'm using v2.0.0 dataset and successfully followed the example on loading Waymo data using dask.

This is all fine for quick testing, but when I use the same method on my data loader  things do not scale so well. Dask is nice but when I actually call Dask's compute() to get the data, it takes sometime even with fast disk.

Data loader when shuffled will sample the frame randomly so I can't have eager loading of each parquet file by looking at its order. 

Example: when it happens that the the  loader samples 10 different frames from 10 different parquet files, then it becomes an I/O bottleneck, even with multiple workers. 

Preloading the whole dataset is out of question due to memory constraint. 

I have been looking at how other frameworks (e.g. Mmdetect) and 3rd party libraries (e.g. Pytorch Waymo loader) are using Waymo: they pre-convert the training frames (e.g. to pickle) so the access time is fast even when frames are randomly sampled.

Is this the recommend way? I feel the  use of parquet file + dask is meant to address this exact issue.

Thanks in advance for the insight.

**Comment by aryasenna** [comment]
In case anyone is wondering:

One possible solution, depending on your use case it to use pushdown filtering.

Shame that the Waymo v2 [example/tutorial](https://github.com/waymo-research/waymo-open-dataset/blob/master/tutorial/tutorial_v2.ipynb) never mentioned the use of Dask's "[pushdown filtering](https://docs.dask.org/en/stable/generated/dask.dataframe.read_parquet.html)."

This should be done when you first read your parquet file:

e.g.

```
image_df = dd.read_parquet(
    os.path.join(directories['CameraImage'], context_name + '.parquet'),
    columns=['key.frame_timestamp_micros', 'key.camera_name', '[CameraImageComponent].image'],
    filters=[('key.frame_timestamp_micros', '==', timestamp), ('key.camera_name', '==', CameraName.FRONT.value)]
)

```
This works for me because my training loader only expects one timestamp and a certain camera. The idea is to make sure you don't load the whole parquet.

I will leave the issue open, or visibility in case the Waymo team wants to update their documentation.

**Comment by JingweiJ** [comment]
Yes push down filtering is a good way for efficiency. We mentioned this a bit in the "A relational database-like structure" section in the [example/tutorial](https://github.com/waymo-research/waymo-open-dataset/blob/master/tutorial/tutorial_v2.ipynb) yet we should discuss more in the aspect of efficiency. Thanks for the advice!

**Comment by JingweiJ** [comment]
Yes push down filtering is a good way for efficiency. We mentioned this a bit in the "A relational database-like structure" section in the [example/tutorial](https://github.com/waymo-research/waymo-open-dataset/blob/master/tutorial/tutorial_v2.ipynb) yet we should discuss more in the aspect of efficiency. Thanks for the advice!

**Comment by aryasenna** [comment]
@JingweiJ Thanks for checking this up. Yes you are correct pushdown filtering mentioned there as a short comment. So technically, it was "mentioned". 

My point being, in the actual  example code where it only uses single frame, it makes sense to use push filtering by default. 🙂

**Comment by nlgranger** [comment]
In case it can help anyone, I have written a library which can load any given data sample from Waymo (and also KITTI, NuScenes or ZOD). As we discussed in https://github.com/waymo-research/waymo-open-dataset/issues/841, one needs to re-encode the parquet files to make random access fast.

The library is here: https://github.com/CEA-LIST/tri3d

It is a bit opinionated because I needed to settle on common conventions across datasets, but I think you'll find it does what you expect it to do most of the time. Notably, it has sane defaults to interpolate poses (ego car, boxes, sensors) such that when you request something at, say, LiDAR frame 12, that something will actually overlap well with the point cloud.
